src/com/mediaengine/rpc/rpc_cdm_mediaengine_handler.cc

```diff
@@ -50,6 +71,170 @@ RpcCdmMediaengineHandler::RpcCdmMediaengineHandler() {
+    goto handle_error;
+  }
+
+  /* Configure a timeout for sendmsg(). This is to avoid an infinite wait. */
+  timeout_tv.tv_sec = SOCKET_SEND_TIMEOUT;
+  timeout_tv.tv_usec = 0;
+  status = setsockopt(socket_fd, SOL_SOCKET, SO_SNDTIMEO, (const void *)&timeout_tv, sizeof(timeout_tv));
+  if(status < 0) {
+    CDM_DLOG() << "connectSocket: Cannot configure sendmsg timeout";
+    goto handle_error;
+  }
+
+handle_error:
+  if(status < 0) {
+    if(socket_fd >= 0) {
+      close(socket_fd);
+      socket_fd = -1;
+    }
+  }
+  return socket_fd;
+}
+
+
+// The caller is responsible to close the returned file descriptor.
+static int allocateIon(size_t size)
+{
+  struct ion_allocation_data alloc_data;
+  struct ion_handle_data hdl_data;
+  struct ion_fd_data fd_data;
+  int ion;
+  int fd = -1;
+
+  ion = open("/dev/ion", O_RDWR);
+  if (ion < 0) {
+    CDM_DLOG() << "Failed to open /dev/ion";
+    return fd;
+  }
+
+  alloc_data.len = size;
+  alloc_data.align = 0;
+  alloc_data.flags = 0;
+  alloc_data.heap_id_mask = 1 << ION_SECURE_HEAP_ID_DECODER;
+
+  if (ioctl(ion, ION_IOC_ALLOC, &alloc_data) == -1) {
+    CDM_DLOG() << "Failed to allocate secure buffer from ION heap ID " << alloc_data.heap_id_mask;
+    goto handle_error;
+  }
+
+  fd_data.handle = alloc_data.handle;
+  if (ioctl(ion, ION_IOC_SHARE, &fd_data) != -1)
+    fd = fd_data.fd;
+  else
+    CDM_DLOG() << "Failed to share secure buffer.";
+
+  hdl_data.handle = alloc_data.handle;
+  (void)ioctl(ion, ION_IOC_FREE, &hdl_data);
+handle_error:
+  close(ion);
+  return fd;
+}
+
+
+static int sendFileDescriptor(int socket_fd, int secure_fd, uint32_t secure_size)
+{
+  int status = -1;
+
+  struct msghdr msg;
+  struct iovec iov;
+
+  /* Control message buffer contains the control message structure plus
+   one file descriptor. */
+  #define CMSG_SIZE (sizeof(struct cmsghdr) + sizeof(int))
+  uint8_t cmsg_buffer[CMSG_SIZE] = {0};
+  struct cmsghdr *cmsg; // Pointer to control message
+
+  if(socket_fd < 0) {
+    CDM_DLOG() << "Invalid socket file descriptor";
+    return -1;
+  }
+
+  /* Send secure buffer size with the file descriptor */
+  iov.iov_base = &secure_size;
+  iov.iov_len  = sizeof(secure_size);
+
+  msg.msg_name       = NULL;
+  msg.msg_namelen    = 0;
+  msg.msg_iov        = &iov;
+  msg.msg_iovlen     = 1;
+  msg.msg_control    = cmsg_buffer;
+  msg.msg_controllen = CMSG_SIZE;
+  msg.msg_flags      = 0; /* ignored */
+
+  cmsg = CMSG_FIRSTHDR(&msg);
+  cmsg->cmsg_len = CMSG_SIZE;
+  cmsg->cmsg_level = SOL_SOCKET;
+  cmsg->cmsg_type = SCM_RIGHTS;
+  *(int *)CMSG_DATA(cmsg) = secure_fd;
+
+  status = sendmsg(socket_fd, &msg, 0);
+  if(status >= 0) {
+    status = 0;  // Success
+  } else {
+    if(errno == EWOULDBLOCK || errno == EAGAIN)
+      CDM_DLOG() << "Cannot send FD (Timeout)";
+    else
+      CDM_DLOG() << "Cannot send FD";
+  }
+
+  return status;
+}
+
+
+static int copySecureData(int secure_fd, uint32_t secure_size, uint8_t *dst, uint32_t dst_size)
```

Comment:
A comment would be good here to highlight that this should only be done as part of the protoype, and relies on memory protection being disabled of the secure buffers.